js/interface/add-markers.js

```diff
@@ -0,0 +1,393 @@
+        const markerData = _.find(this.widgetData.markers, { name: marker[this.markerTypeColumn] })
+        return { 
+          id: marker.id,
+          name: marker[this.markerNameColumn],
+          floor: marker[this.markerFloorColumn],
+          type: marker[this.markerTypeColumn],
+          icon: markerData ? markerData.icon : '',
+          color: markerData ? markerData.color : '#333333',
+          positionx: marker[this.markerXPositionColumn],
+          positiony: marker[this.markerYPositionColumn],
+          updateName: false,
+          copyOfName: ''
+        }
+      })
+
+      return newMarkerData
+    },
+    setActiveMarker(index) {
+      if (this.activeMarker !== index) {
+        this.activeMarker = index
+        this.setupPinchZoomer()
+      }
+    },
+    updateFloor(floorName, index) {
+      this.mappedMarkerData[index].floor = floorName
+      this.setupPinchZoomer()
+    },
+    updateMarker(marker, index) {
+      this.mappedMarkerData[index].type = marker.name
+      this.mappedMarkerData[index].icon = marker.icon
+      this.mappedMarkerData[index].color = marker.color
+      this.setupPinchZoomer()
+    },
+    toggleUpdateName(index, currentName) {
+      const $vm = this
+      this.mappedMarkerData[index].updateName = !this.mappedMarkerData[index].updateName
+
+      if (currentName) {
+        this.mappedMarkerData[index].copyOfName = currentName
+        this.$nextTick(() => $vm.$refs['changename-' + index][0].focus())
+      }
+    },
+    cancelNameUpdate(index) {
+      this.mappedMarkerData[index].name = this.mappedMarkerData[index].copyOfName
+      this.mappedMarkerData[index].copyOfName = ''
+      this.toggleUpdateName(index)
+    },
+    deleteMarker(index) {
+      this.mappedMarkerData.splice(index, 1)
+    },
+    nameWithId({ name, id }) {
+      return `${name} — [${id}]`
+    },
+    createNewData() {
+      const $vm = this
+      const name = `${this.appName} - Markers`
+      Fliplet.Modal.prompt({
+        title: 'Please type a name for your data source:',
+        value: name
+      }).then((name) => {
+        if (name === null || name === '') {
+          return Promise.reject()
+        }
+
+        return name
+      }).then((name) => {
+        return Fliplet.DataSources.create({
+          name: name,
+          organizationId: organizationId
+        })
+      }).then((ds) => {
+        $vm.dataSources.push(ds)
+        $vm.selectedDataSource = ds.id
+      })
+    },
+    chooseExistingData() {
+      const $vm = this
+      Fliplet.Modal.confirm({
+        title: 'Changing data source',
+        message: '<p>If you continue the data source we created for you will be deleted.</p><p>Are you sure you want to continue?</p>'
+      }).then((result) => {
+        if (!result) {
+          return
+        }
+
+        return $vm.deleteDataSource()
+      }).then(() => {
+        // Remove from dataSources
+        $vm.dataSources = _.filter($vm.dataSources, (ds) => {
```

Comment:
You have used an arrow function here, so there's no need to capture `$vm` in a closure, and just use `this` here.
```suggestion
        this.dataSources = _.filter(this.dataSources, (ds) => {
```

Reply:
@angelnikolov You, if you are going to highlight every time I use `$vm` unnecessarily, you will have to write lots of comments.

I'll make it easy for you, let me know the basics of when do I need to set a `const $vm = this` ?

That way I can use that knowledge and review my code myself and change where needed.

Reply:
I can't find the link I sent you on Slack, but that sums it up perfectly.
Basically, when you have an arrow function like this `() => {}` it doesn't create a new context so in a sense it uses or inherits the parent one. This is useful, when you don't care about the context of invocation of the function you are currently in, i.e in your case - the first callback of the `Promise.then` method.
So in your example, the context will be the Vue object you are building, which has all your methods and properties, so you could directly use them.
I suggest that you just remove all $vm and change them with `this`, see if everything still works and just roll with `this` if it does :)